### PR TITLE
Several QoR and minor generator functionality improvements for Rocket-related utilities

### PR DESCRIPTION
**Pull request by aswaterman**
In trying to reuse some of these units for another generator I'm working on,  I noticed some deficiencies in QoR and also wanted a few more knobs to turn.

### CODE REVIEW

File: src/main/scala/rocket/DCache.scala [hunk 1]
```diff
@@ -449,19 +459,24 @@ class DCacheModule(outer: DCache) extends HellaCacheModule(outer) {
...
+    val uncachedGrantOpcodes = uncachedGrantOpcodesWithData ++ uncachedGrantOpcodesSansData
+    val whole_opc = tl_out.d.bits.opcode
     if (usingDataScratchpad) {
-      assert(!(tl_out.d.valid && res))
-      false.B
+      assert(!tl_out.d.valid || whole_opc.isOneOf(uncachedGrantOpcodes))
+      // the only valid TL-D messages are uncached, so we can do some pruning
+      val opc = whole_opc(uncachedGrantOpcodes.map(_.getWidth).max - 1, 0)
+      val data = DecodeLogic(opc, uncachedGrantOpcodesWithData, uncachedGrantOpcodesSansData)
+      (opc, true.B, data)
```
Comment (hcook) [inline src/main/scala/rocket/DCache.scala]: Seems like all of this should get sucked into some kind of helper method on `TLMessages`?

Comment (aswaterman) [reply src/main/scala/rocket/DCache.scala]: Per in-person discussion, postponing until TL message-ID mapping is revisited and associated code is refactored.

File: src/main/scala/rocket/TLB.scala [hunk 2]
```diff
@@ -79,31 +82,92 @@ class TLB(instruction: Boolean, lgMaxSize: Int, nEntries: Int)(implicit edge: TL
...
   }
 
-  val totalEntries = nEntries + 1
-  val normalEntries = nEntries
-  val specialEntry = nEntries
-  val aeEntry = specialEntry - (1 << log2Floor(nEntries))
-  val valid = Reg(init = UInt(0, totalEntries))
-  val reg_entries = Reg(Vec(totalEntries, UInt(width = new Entry().getWidth)))
-  val entries = reg_entries.map(_.asTypeOf(new Entry))
+  class Entry(nSectors: Int, superpageOnly: Boolean) extends Bundle {
```
Comment (hcook) [inline src/main/scala/rocket/TLB.scala]: FYI if you turn these into `val`s you shouldn't need to define your own `cloneType` anymore

Comment (aswaterman) [reply src/main/scala/rocket/TLB.scala]: awesome

**Pull request merged**